runtime/memory.c

```diff
@@ -691,37 +687,90 @@ CAMLexport void caml_local_region_end(intnat reg)
   Caml_state->local_sp = reg;
 }
 
-//#define Local_init_wsz 64
-#define Local_init_wsz (4096)
-void caml_local_realloc()
+CAMLexport caml_local_arenas* caml_get_local_arenas()
 {
-  intnat new_bsize;
-  struct region_stack* stk;
-  char* stkbase;
-  if (Caml_state->local_top == NULL) {
-    new_bsize = Bsize_wsize(Local_init_wsz);
+  caml_local_arenas* s = Caml_state->local_arenas;
+  if (s != NULL)
+    s->saved_sp = Caml_state->local_sp;
+  return s;
+}
+
+CAMLexport void caml_set_local_arenas(caml_local_arenas* s)
+{
+  Caml_state->local_arenas = s;
+  if (s != NULL) {
+    struct caml_local_arena a = s->arenas[s->count - 1];
+    Caml_state->local_sp = s->saved_sp;
+    Caml_state->local_top = (void*)(a.base + a.length);
+    Caml_state->local_limit = - a.length;
   } else {
-    CAMLassert((char*)Caml_state->local_top + Caml_state->local_limit == Caml_state->local_top->base);
-    new_bsize = -Caml_state->local_limit;
+    Caml_state->local_sp = 0;
+    Caml_state->local_top = NULL;
+    Caml_state->local_limit = 0;
   }
-  while (Caml_state->local_sp < -new_bsize) new_bsize *= 2;
-  stkbase = caml_stat_alloc(new_bsize + sizeof(struct region_stack));
-  stk = (struct region_stack*)(stkbase + new_bsize);
-  memset(stkbase, 0x42, new_bsize); /* FIXME debugging only */
-  stk->base = stkbase;
-  stk->next = Caml_state->local_top;
-  Caml_state->local_top = stk;
-  Caml_state->local_limit = -new_bsize;
+}
+
+void caml_local_realloc()
+{
+  caml_local_arenas* s = caml_get_local_arenas();
+  intnat i;
+  char* arena;
+  if (s == NULL) {
+    s = caml_stat_alloc(sizeof(*s));
+    s->count = 0;
+    s->next_length = 0;
+    s->saved_sp = Caml_state->local_sp;
+  }
+  if (s->count == Max_local_arenas)
+    caml_fatal_error("Local allocation stack overflow - exceeded Max_local_arenas");
+
+  do {
+    if (s->next_length == 0) {
+      s->next_length = Init_local_arena_bsize;
+    } else {
+      /* overflow check */
+      CAML_STATIC_ASSERT(((intnat)Init_local_arena_bsize << (2*Max_local_arenas)) > 0);
+      s->next_length *= 4;
+    }
+    /* may need to loop, if a very large allocation was requested */
+  } while (s->saved_sp + s->next_length < 0);
+
+  arena = caml_stat_alloc_noexc(s->next_length);
+  if (arena == NULL)
+    caml_fatal_error("Local allocation stack overflow - out of memory");
+  for (i = 0; i < Wsize_bsize(s->next_length); i++) {
+    ((header_t*)arena)[i] = Local_uninit_hd;
```

Comment:
Is this initialisation here for debugging reasons or is it required for correctness?

Reply:
This is required for correctness. (The debugging-related inits are guarded by `#ifdef DEBUG`, and fill with `Debug_foo`)